### DISCUSSION TITLE
Switch to "Identity" columns instead of "Serial" for PostgreSQL 10+

### DISCUSSION THREAD

**Discussion by mboeringa** [osm2pgsql-dev/osm2pgsql#1538] [Ideas]
PostgreSQL 10 introduced the "Identity" column constraint for unique ID columns with SQL standard definition ("GENERATED ALWAYS / BY DEFAULT AS IDENTITY"), which has several advantages over the "serial" type. Maybe it would be an idea to at least allow specifying this sql_type instead of serial, or default to it for PostgreSQL 10+:

https://www.depesz.com/2017/04/10/waiting-for-postgresql-10-identity-columns/

https://www.postgresql.org/docs/current/sql-createtable.html

**Comment by joto** [comment]
You can use any `sql_type` you like, this is the point of the `sql_type` setting. I am putting something in the manual that this is similar to a serial column.

**Reply by mboeringa** [reply]
"Identity" is actually a bit different than "serial", in that "serial" defines both a column type and column constraint with a single keyword, while for identity, the column type would be something like "int" or "bigint", and you add the column constraint by adding the "GENERATED ALWAYS / BY DEFAULT AS IDENTITY" after the specification of the column type.

But if I understand you well, you are saying that if I specify

`column = 'ID', sql_type = 'bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY'`

in the Lua style file, this would already work in the current implementation?

**Reply by joto** [reply]
Yes, i believe it should work, but we will not guarantee that it will always work.

**Reply by mboeringa** [reply]
Ok, I will try it out and see if I can get it working, and report back the result here.

**Reply by mboeringa** [reply]
@joto 

You are right, this is already supported.

My first attempt failed though:

![afbeelding](https://user-images.githubusercontent.com/7635726/125176254-5d527980-e1d2-11eb-8e07-980bcd7961d9.png)

But I then realized that the code should read:

`column = 'ID', sql_type = 'bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY', create_only = true`

to prevent the not-null constraint error, as properly documented on the manual pages.

After changing that, both 'serial' and the 'identity' clauses worked.

I did realize there is another reason good reason to switch to 'identity' though: I noticed in DBeaver it appears the 'serial' column type is created as 'int4' by default by PostgreSQL. While not an immediate concern, it is clear though that eventually one may run out of values with 'int4' and the ever growing OpenStreetMap database... It is not without reason osm_id has been 'int8' for quite some time.

**Reply by rustprooflabs** [reply]
> I did realize there is another reason good reason to switch to 'identity' though: I noticed in DBeaver it appears the 'serial' column type is created as 'int4' by default by PostgreSQL. While not an immediate concern, it is clear though that eventually one may run out of values with 'int4' and the ever growing OpenStreetMap database... It is not without reason osm_id has been 'int8' for quite some time.

@mboeringa Using `bigserial` instead of `serial` will create `int8`. `IDENTITY` is still the preferred method though, it conforms to the SQL standard and makes permissions easier to manage.

**Reply by mboeringa** [reply]
@rustprooflabs, thanks. As I showed and Jochen stated, "identity" is already usable with current v1.5.0 of osm2pgsql, so that is good to know for style developers.
